src/main/java/com/aws/greengrass/deployment/ConnectivityValidator.java

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.deployment;
+
+import com.aws.greengrass.config.Topic;
+import com.aws.greengrass.config.Topics;
+import com.aws.greengrass.deployment.errorcode.DeploymentErrorCode;
+import com.aws.greengrass.deployment.exceptions.ComponentConfigurationValidationException;
+import com.aws.greengrass.deployment.model.DeploymentResult;
+import com.aws.greengrass.logging.api.Logger;
+import com.aws.greengrass.logging.impl.LogManager;
+import com.aws.greengrass.mqttclient.MqttClient;
+import com.aws.greengrass.util.Coerce;
+import com.aws.greengrass.util.GreengrassServiceClientFactory;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import software.amazon.awssdk.crt.mqtt.MqttClientConnection;
+import software.amazon.awssdk.crt.mqtt.MqttException;
+import software.amazon.awssdk.http.SdkHttpClient;
+import software.amazon.awssdk.iot.AwsIotMqttConnectionBuilder;
+import software.amazon.awssdk.services.greengrassv2data.GreengrassV2DataClient;
+import software.amazon.awssdk.services.greengrassv2data.model.ListThingGroupsForCoreDeviceRequest;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class ConnectivityValidator implements AutoCloseable {
+    private static final Logger logger = LogManager.getLogger(ConnectivityValidator.class);
+    private static final int maxRetries = 3;
+    public static final String clientIdSuffix = "#validation";
+    private static final int mqttWaitTimeSeconds = 2;
+    private static final int httpWaitTimeSeconds = 5;
+    private final DeviceConfiguration deploymentConfiguration;
+    private final MqttClient mqttClient;
+    private final GreengrassServiceClientFactory factory;
+
+    private AwsIotMqttConnectionBuilder builder;
+    private MqttClientConnection connection;
+    private SdkHttpClient httpClient;
+    private GreengrassV2DataClient greengrassV2DataClient;
+
+
+    /**
+     * Constructor.
+     *
+     * @param mqttClient       Used to create AWS Mqtt Connection Client
+     * @param factory          Used by integration tests to set mock
+     * @param currentConfig    current device configs
+     * @param deploymentConfig Map of the configs to be merged in and validated against
+     * @param timestamp        time stamp of the deployment
+     */
+    public ConnectivityValidator(MqttClient mqttClient, GreengrassServiceClientFactory factory,
+                                 DeviceConfiguration currentConfig, Map<String, Object> deploymentConfig,
+                                 long timestamp) {
+        this.mqttClient = mqttClient;
+        this.factory = factory;
+        deploymentConfiguration = new DeviceConfiguration(currentConfig, deploymentConfig, timestamp);
```

Comment:
why are you making the device config based on the _currentConfig_? It should be based on the new config to validate right?

Reply:
The current config is copied over to preserve time stamps, then the new deployment config will be merged in on top of that using the deployment timestamp

Reply:
Where is this merging? It should not be merging in device configuration. Pass on the actual final configuration to device config

Reply:
Sorry could have worded it better.

In the kernel constructor I added (which I plan to remove since we shouldn't create a new kernel or context)
I create a new config and use `copyFrom` on the current config to preserve name / value / modTime, then I `mergeMap` the deployment config using the deployment timestamp